helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/HardConstraint.java

```diff
@@ -29,11 +29,39 @@
  */
 abstract class HardConstraint {
 
+  /**
+   * Result of the constraint validation, which contains the error message if the validation fails.
+   */
+  static class ValidationResult {
+
+    public static final ValidationResult OK = new ValidationResult(true, null);
+
+    private final boolean success;
+    private final String errorMsg;
+
+    public static ValidationResult fail(String errorMsg) {
+      return new ValidationResult(false, errorMsg);
+    }
+
+    private ValidationResult(boolean success, String errorMsg) {
+      this.success = success;
+      this.errorMsg = errorMsg;
+    }
+
+    public String getErrorMsg() {
+      return errorMsg;
+    }
+
+    public boolean isSuccess() {
+      return success;
+    }
+  }
+
```

Comment:
That's too much to have a nested class here. Why not just let it be part of HardConstraint property instead of creating a nested class.

Reply:
Thanks for the review Junkai. Yes, I agree that nested class does not look good here. I added here for the reason that its purpose and use is tied within the scope of HardConstraint validation. 

So, we can have error message as part of  HardConstraint, but currently we have only one object of each hard constraint, which is used everywhere (maintained in ConstraintBasedAlgorithmFactory). This will be possible if we create a constraint object per assignable replica and node combination.